### ISSUE TITLE
CSV cannot append CSV with column names that truncate to the same prefix

### ISSUE THREAD

**Issue by crisptrutski** [metabase/metabase#44725]
### Describe the bug

Many databases have modest limits in the length of the column names that they support, and they may truncate them silently. When this truncation would result in two columns having the same name, ~~the upload~~ appending will fail.

**Update:** We have updated the "unique-ification" of columns to keep its deduplicated names within the maximum length. This fixed initial uploads, but appends are still broken as we do not modify them to ensure uniqueness.

### To Reproduce

1. Enable uploads for a Postgres database.
2. Go to a collection.
3. Upload a CSV file with two columns that differ only in their 64th character.
4. ~~See the following error:~~ *This now works for initial upload.*
5. Append the same file again.
6. See the following error:

> Error executing write query: ERROR: column "really_long_name_with_a_difference_that_will_only_appear_later_o" specified more than once

### Expected behavior

~~The upload succeeds and there are two distinct columns in the final table.~~

Appending to the relevant model will succeed.

### Information about your Metabase installation

```JSON
- Metabase version: 48+
```

### Severity

P2

**Comment by crisptrutski** [comment]
A solution to this is to explicitly truncate names shorter than necessary, so that we can append some disambiguating identifier at the end, e.g. a counter.

Unfortunately we use the database column name verbatim in the model, and these suffixes will have no meaning to the customer, who will thus have difficulty telling them apart. Additionally, this would make additional uploaded sensitive to column ordering.

A solution to the previous two issues would be to store the non-truncated (and potentially non-munged) name somewhere else in the database, and to use *this* name for both the labels in the model and matching up with appended files (see https://github.com/metabase/metabase/issues/44724)

**Comment by crisptrutski** [comment]
So, we can theoretically use `metabase_field.displace_name` to store the original name, which would be spark joy in and of itself. Unfortunately, that is also bounded, which is two characters shorter than h2's identifier limit. For Postgres this would still give us an additional ±200 characters though.

I'm guessing the reason we don't unique-ify column names on append, is because we could switch up columns due to different ordering. Perhaps @calherries can clarify this. If it is the case, the "does this look right" confirmation screen idea could solve the problem.

A probably-too-smart-for-its-own-good idea would be to use a hash of the full name as a suffix of the physical column name, but that could still have collisions. A possibly overly heavy solution would be to introduce a new table joining to the field, with a text column. We could only populate it for the long ones, although migrating from h2 to postgres would change the definition (and result in missing values).

Probably 254 characters is ample, since humans have to read these labels after all. If we can restrict this issue to only affecting longer titles, that is probably enough.

**Comment by calherries** [comment]
Having a look at the solution in https://github.com/metabase/metabase/pull/44809 I'm not convinced this issue is important enough to justify the extra complexity and magic. It's already feels a little too magic that we de-duplicate field names or truncate long names and it's even more magic to de-duplicate and truncate long names to the exact number of characters required to store the required suffix to de-duplicate them. To demonstrate what I mean by complexity, imagine explaining this behaviour to a customer?
```
(upload.u/uniquify-names 4 ["abcd" "abcd" "ab_2" "ab_3"])
;; => ("abcd" "ab_2" "ab_3" "ab_4")
```
Without any other bright ideas I suggest we just keep it simple and throw an exception if there are long column names that truncate to the same prefix. To quote [Arakaki](https://metaboat.slack.com/archives/C0641E4PB9B/p1719406719386459?thread_ts=1719378119.051019&cid=C0641E4PB9B):  "This is the kind of issue that is between a bug and a limitation. I’d say that it is reasonable to expect not huge column names."

I liked the idea of using the display name to identify fields with duplicate slugified names, but I imagine most of its benefit would be from identifying names whose uniqueness had been removed by slugifying non-ASCII characters. The benefit is not for truncated long names: that case seems a lot rarer in practice.

**Comment by crisptrutski** [comment]
I agree that "usurping" prefixes is confusing. It would be even worse in the case `["abcd" "abcd" "ab_2"]`, where we steal a name verbatim. That said, a surprising name is better than surprising behavior in my opinion. Coupled with using the original names as labels, and querying through models, I still lean in the direction of making this work. If you look at my [PR](https://github.com/metabase/metabase/pull/44873/files) which removes the buffer, and therefore introduces duplicates, we still have other problems to resolve.

That said, I wholeheartedly agree with fixing appends for existing tables, and we'd need to get that fix into the deployment going out on Tuesday, so happy to put this on the backburner.

> It's far too much magic to de-duplicate and truncate long names to the exact number of characters required to store the required suffix to de-duplicate them

This is something the original utility calls out support for, so I don't think it's that outlandish an idea, and I also think we can make it work. I've updated the PR with a simple modification that ensures that we reserve a given name for the first column whose prefix matches it verbatim, to prevent any insolent usurper strings flipping things around rudely.

```clojure
(#'metabase.upload/derive-column-names ::test-driver ["abcd" "abcd" "ab_2_bc" "ab_3_bc"])
;; => (:abcd :ab_4 :ab_2 :ab_3)
```

**Comment by crisptrutski** [comment]
Update: I see actually MBQL thought about this edge case, and its solution is to resort to hashes for really long names. That's much less likely to result in these "usurping" gotchas, and also saves us forking this big method, so going to put up a PR just using that next.

**Comment by crisptrutski** [comment]
Urgh, meant to close the PR not the issue 🤦

**Comment by luizarakaki** [comment]
Moving this to P3 while we don't hear about people hitting this.
There is also a workaround (use shorter column headers)
